src/instruction.rs

```diff
@@ -213,143 +213,239 @@ impl fmt::Display for MemoryReference {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+pub struct Gate {
+    pub name: String,
+    pub parameters: Vec<Expression>,
+    pub qubits: Vec<Qubit>,
+    pub modifiers: Vec<GateModifier>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CircuitDefinition {
+    pub name: String,
+    pub parameters: Vec<String>,
+    // These cannot be fixed qubits and thus are not typed as `Qubit`
+    pub qubit_variables: Vec<String>,
+    pub instructions: Vec<Instruction>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct GateDefinition {
+    pub name: String,
+    pub parameters: Vec<String>,
+    pub matrix: Vec<Vec<Expression>>,
+    pub r#type: GateType,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Declaration {
+    pub name: String,
+    pub size: Vector,
+    pub sharing: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Measurement {
+    pub qubit: Qubit,
+    pub target: Option<MemoryReference>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Reset {
+    pub qubit: Option<Qubit>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CalibrationDefinition(pub Calibration);
```

Comment:
Agreeing with what @dbanty wrote below; this does not need a double nesting. The reason it was like this was because I needed the functionality which we're extending here to all instructions; I needed to be able to pull out the contents and use them independently from `Instruction`.

`Calibration` can just become `CalibrationDefinition` here.